rfc/0002-report-mode.md

```diff
@@ -0,0 +1,122 @@
+> closely to the solution you have > in mind. If this RFC is not accepted, the motivation can be
+> used to develop alternative solutions.
+
+Currently, Policies are deployed and become active immediately -- accepting, mutating or rejecting
+requests.
+
+With this context, as a cluster administrator it is not easy to understand the consequences of
+deploying new policies in a running cluster. This RFC works on the strategy to improve this
+situation and let the administrator gain visibility before policies are truly enforced in the
+cluster.
+
+Introduce a report mode in Kubewarden that acts in the following way:
+
+1. Deploy policies without rejecting or mutating requests, so actions that would have been taken had
+   the policy been active are logged. We will call this method "reporting"  mode, because it lets
+   the cluster operator understand if a policy in this mode would be rejecting requests that impact
+   the regular functionality of the cluster.
+
+## Examples / User Stories
+[examples]: #examples
+
+> Examples of how the feature will be used. Interactions should show the action and the response.
+> When appropriate, provide user stories in the form of "As a [role], I want [feature], so [that]."
+
+As a Kubernetes administrator I want to deploy new policies in a state that they don't perform any
+request rejections, but only inform about what would have been rejected had the policy been active.
+
+As a Kubernetes administrator I want to deploy new policies in "reporting" mode, and then flip the
+switch of the policy to activate it after it has been in "reporting" mode for a short while that I
+considered enough to learn if it would have an undesired impact in the cluster.
+
+As a Kubernetes administrator I want to plot in Grafana acceptions/rejections/mutations coming from
+policies in "report" mode in the same way that I do with policies in "enforce" mode -- the only mode
+Kubewarden had until now.
+
+# Detailed design
+[design]: #detailed-design
+
+> This is the bulk of the RFC. Explain the design in enough detail for somebody familiar with the
+> product to understand, and for somebody familiar with the internals to implement.
+
+> This section should cover architecture aspects and the rationale behind disruptive technical
+> decisions (when applicable), as well as corner-cases and warnings.
+
+## "report" mode
+
+### CRD
+
+The `ClusterAdmissionPolicy` CRD will be extended so it includes a `mode` attribute:
+
+```go
+type ClusterAdmissionPolicySpec struct {
+   // ...
+   Mode string `json:"mode,omitempty"`
+   // ...
+}
+```
+
+Two values are allowed: `enforce` and `report`. If `mode` is omitted, it will be defaulted to
+`enforce`. It's possible to update the `mode` from `report` to `enforce`. Allowing to change `mode`
+from `enforce` to `report` requires further consideration -- so it's not possible to deactivate
+policies just by having `UPDATE` permissions.
```

Comment:
I don't understand what would prevent us from changing a policy mode from "enforce" to "report". You hint at a technical limitation, what is that about?

Reply:
RBAC won't help us here. I think the only way we can prevent this is by using our own webhook to disable `enforce` -> `report` transition for security. If someone wants to disable a policy temporarily they need to have DELETE permissions, and then they can recreate with `report` if desired. Point is, if a user only has `UPDATE` and `CREATE` permissions, they should not be allowed to transition from `enforce` to `report`.

Reply:
Since transitioning from `report` to `enforce` is stronger, I don't see a reason to forbid.

Reply:
>  Since transitioning from report to enforce is stronger, I don't see a reason to forbid.

It could produce DOS against something not properly configured for example, I would forbid that change too.

Reply:
Note that what I am proposing is going to stronger is fine. Not the other way around. If that causes a DOS is not our fault IMO, we have to provide the tools for that to not happen (like the report mode)

Reply:
For what is worth they could have DOS'ed from the very beginning by creating the policy in enforce mode directly.

Reply:
I see, thanks for the explanations.

Overall I agree with:
* `report` -> `enforce` being allowed :green_circle: 
* `enforce` -> `report` **not** being allowed :red_circle: